soc: arm: st_stm32: Cleanup gpio function code.

cybertale opened the pull request:
f0/f2/f3/f4/f7/l0/l4 series STM32 have mostly the same GPIO architecture
and can share the same code for GPIO manipulation. Only f1 series chips
have standalone architecture.
Functions of the external interrupt line control are also the same.
This patch extracts common code from them and put them into the 'common'
folder.

Functions of control GPIO of these series scattered in
soc/arm/st_stm32/stm32xx/ folders contain these functions:
stm32_gpio_flags_to_conf(), stm32_gpio_configure(), stm32_gpio_set(),
stm32_gpio_get, stm32_gpio_enable_int().
This patch move them into the soc_gpio-no-f1.c file and add them as
sources when these series chips are selected.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale commented:
I tested this code on a stm32f407 based custom board.

erwango reviewed (commented):
@cybertale, this was in my TODO list for a while, but never got the time to do it. This is a welcome change, thanks!

As a first feedback, can you look into including F1 in the change?
It will be easier if everything sits at the same place, and there is mainly 1 really divergent function (stm32_gpio_configure), this could be handled in the same file. stm32_gpio_flags_to_conf as similar structure and it could be tricked to be compatible.

Then, if you're ready for it, a nice addition would be to use STM32Cube LL API to even reduce the code base:
  - In function stm32_gpio_enable_int, use LL_SYSCFG_SetEXTISource/LL_GPIO_AF_SetEXTISource 
  - In stm32_gpio_set, use LL_GPIO_SetOutputPin/LL_GPIO_ResetOutputPin
  - In stm32_gpio_configure, LL_GPIO_SetPinMode, LL_GPIO_SetPinOutputType, ..
  - ...

cybertale commented:
@erwango 
Hi erwango, 
Strange I did look into this thread yesterday but didn't see your reply, sorry for the late reply!
I'm very familiar with the F1 series chips, I'll see what I can do.

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	if (direction == GPIO_DIR_OUT) {
+#ifdef CONFIG_SOC_SERIES_STM32F1X
```
It would be clearer to define a smart set of headers and remove all this ifdefery from this function
```suggestion
#ifdef CONFIG_SOC_SERIES_STM32F1X
#define PIN_MODE_OUTPUT STM32_MODE_OUTPUT | STM32_CNF_GP_OUTPUT | STM32_CNF_PUSH_PULL
#define PIN_MODE_INPUT ...
#define PIN_PUPD_PULLUP ...
...
#else
#define PIN_MODE_OUTPUT   STM32_MODER_OUTPUT_MODE
#define PIN_MODE_INPUT ...
#define PIN_PUPD_PULLUP ...
```

cybertale replied:
I just finished implementing these functions with LL library. 
I was just want to ask should I push only one commit using the LL library or two commits that the first one merge separate soc_gpio.c to one file, and the second commit change the implementation to use the LL library because one commit should do only one thing?

cybertale replied:
I have a f103zet6 based board and a f407zgt6 based board. Usually I test on both of them once I make some changes.

erwango replied:
> I just finished implementing these functions with LL library.
I was just want to ask should I push only one commit using the LL library or two commits that the first one merge separate soc_gpio.c to one file

Ok, great thanks. Please split into 2 commits, it will be easier to review.

erwango replied:
>  f103zet6 based board and a f407zgt6

Ok, thanks. I'll do some additional tests on the boards I own.

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,271 @@
+ * @brief
+ *
+ * Common manipulation functions for GPIO of
+ * stm32f0/f1/f2/f3/f4/f7/l0/l4 series.
```
Update to "... GPIO of STM32 series"

erwango reviewed (changes requested):
Few comments otherwise sounds good.
Adding LL would have been nice but can be done latter on, so is not blocking.

On which series have you been able to test this?

cybertale commented:
Hi @erwango, I updated the code, how about this time?

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+/* F1 series STM32 use AFIO register to configure interrupt generation
+ * and pinmux.
+ */
+#ifdef CONFIG_SOC_SERIES_STM32F1X
```
Shouldn't this be removed in the LL commit?

cybertale replied:
Yes, and should syscfg_registers.h be removed?? Since we prefer using the LL library so they may will not be used in the future.

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+#include <gpio.h>
+#include <gpio/gpio_stm32.h>
+#include <soc.h>
+#include "soc_syscfg_common.h"
```
Is that still required in LL commit?

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+	return 0;
+}
+
+int stm32_gpio_set(u32_t *base, int pin, int value)
```
Could be moved in stm32 gpio driver (merged in gpio_stm32_write)

cybertale replied:
So I'm thinking all functions in soc_gpio.c file can be moved into gpio_stm32.c file. Should moving them into gpio_stm32.c be a separate commit ?

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+
+int stm32_gpio_get(u32_t *base, int pin)
+{
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)base;
```
Could be moved in stm32 gpio driver (merged in gpio_stm32_read)

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+	return (LL_GPIO_ReadInputPort(gpio) >> pin) & 0x1;
+}
+
+int stm32_gpio_enable_int(int port, int pin)
```
Move to gpio driver. Could remain a separate function.

erwango commented on soc/arm/st_stm32/common/soc_syscfg_common.h:
```diff
@@ -0,0 +1,18 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __SOC_SYSCFG_COMMON_H
```
Is that file still required?

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+	pin = stm32_f1_pinval_get(pin);
+
+	u32_t temp = conf & (STM32_MODE_INOUT_MASK << STM32_MODE_INOUT_SHIFT);
+	if (temp == STM32_MODE_INPUT) {
```
Break a line before "if..." for readability.

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(altf);
+
+	pin = stm32_f1_pinval_get(pin);
```
Put this function common to all socs.

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+#endif
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+static inline u32_t stm32_f1_pinval_get(int pin)
```
Get this function common to all SoCs.
Put  #îfdef internal to the function

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+
+	} else {
+		temp = conf & (STM32_CNF_OUT_1_MASK << STM32_CNF_OUT_1_SHIFT);
+		if (temp == STM32_CNF_GP_OUTPUT) {
```
Break a line be for if

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+	pin = 1 << pin;
+
+	u32_t temp = conf & (STM32_MODER_MASK << STM32_MODER_SHIFT);
+	if (temp == STM32_MODER_OUTPUT_MODE) {
```
Break a line before if

erwango commented on soc/arm/st_stm32/common/soc_gpio.c:
```diff
@@ -0,0 +1,272 @@
+	ARG_UNUSED(altf);
+
+	pin = stm32_f1_pinval_get(pin);
+
```
Add some comments to this function.

erwango reviewed (commented):
Here are some clues to make code more readable.

dbkinder commented:
OK, I removed me as a reviewer.

cybertale commented:
Hi @erwango ,
Could you offer some debug information if you got some time?
Here are my debug procedure. I believe you understand this better than me, I'm just ensure that we are using the same commands and see the same stuff.

In build directory of samples/basic/disco, after run 'ninja' and 'ninja flash'
run 
On F1 platform, run
`ninja debug`
`b soc_gpio.c:214`
`p gpio`
`p gpio->CRH`
`p gpio->CRL`
`c`
`p gpio`
`p gpio->CRH`
`p gpio->CRL`
On F7 platform, run
`ninja debug`
`b soc_gpio.c:214`
`p gpio`
`p gpio->MODER`
`c`
`p gpio`
`p gpio->MODER`

And of course the LEDs ports and pin numbers of your boards or what your boards are and I can look it up myself. 
Running into breakpoints twice because LEDs of some boards are in different port.
Though I don't have a F7 board(actually I have one, just got it lost in the lab...), I think the F4 board I have should act just the same.

erwango commented:
@cybertale , won't have time to do it today sorry. Will have a try tomorrow

cybertale commented:
No hurry.

erwango commented:
Ok, so actually, to work with sample/disco on all boards requires a rebase...
And this is working!

Let me have a look to sample/basic/button

erwango commented:
@psidhu , btw, you might be interested by that one.

cybertale commented:
@erwango Great the disco sample works! My boards are from a Chinese manufacture and there aren't many information from Google. They are Atomic Explorer(F4) and Atomic Warship(F1).
Sorry there seems like is a bug in reading. I got it worked earlier and only tested the disco sample in the latest commit, I'm debugging it now.

cybertale commented:
@erwango Sorry for my mistake, it was because a error in index translation in soc_gpio:252. I've fixed it, give the new code a try if you got some time!

b0661 commented on soc/arm/st_stm32/stm32f0/soc.h:
```diff
@@ -55,6 +55,10 @@
 #include <stm32f0xx_ll_spi.h>
 #endif
 
+#ifdef CONFIG_GPIO_STM32
```
Why not use CONFIG_USE_STM32_LL_GPIO instead?

It was created to flag the usage of LL_GPIO. Using CONFIG_GPIO_STM32 restricts the inclusion to the STM32_GPIO driver - but other drivers may also want to use the LL_GPIO functions.

cybertale replied:
Understand, while I think it's because for other drivers, the interface of enabling gpio driver of stm32 should be always CONFIG_GPIO_STM32. Other drivers like spi drivers should use generic configuration functions to configure the port, pin, and the gpio driver will handle the configuration. No matter if the GPIO driver is using the LL library.
There is another reason for it, other drivers define CONFIG_SPI_STM32, CONFIG_I2C_STM32_V2... 
So, I'm new to zephyr and just expressing what I'm thinking about this problem, and I think this may should be discussed in a new thread.

erwango replied:
@b0661 , USE_STM32_LL_XXX is made for LL sources, not headers.
Using these flags in soc.h would work but it's a new way of using them and this deserve a dedicated discussion.

erwango commented:
@cybertale , thanks for the change.
I've made few tests with samples/basic/button 
* Working on F1 / F4 / F7 / L4
* Not working on F0 / L0 / F3. On F3 UART is not working anymore

Seems that some bug(s?) remain. Can you have a check ?

cybertale commented:
@erwango Strange, I checked with the reference manuals, F0/L0/F3 and F4/F7 have almost identical GPIO and SYSCFG->EXTICR[1-4] architecture. Only in F0/L0 series alternate functions are limited to 8 kind, and some initial values of MODER registers are different. 
For UART they are also the same. Currently I do not have any boards with F0/L0/F3 series of chips. 
Could you have a little debug on them if you got some time?
`ninja debug`
`b soc_gpio.c:106`
`p gpio`
`p gpio->MODER`
`p gpio->PUPDR`
print the layout of EXTICRs
`x 0x40010008`
`x 0x4001000c`
`x 0x40010010`
`x 0x40010014`

erwango commented:
@cybertale 

Debug on nucleo_l073rz.
```
Breakpoint 1, stm32_gpio_configure (base_addr=base_addr@entry=0x50000000, pin=pin@entry=2, conf=conf@entry=548, altf=altf@entry=4) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:149
149		pin = 1 << pin;
(gdb) p gpio
$1 = (GPIO_TypeDef *) 0x50000000
(gdb) p pin
$2 = 2
(gdb) p /x (GPIO_TypeDef)*gpio
$3 = {MODER = 0xebfffcff, OTYPER = 0x0, OSPEEDR = 0xc000000, PUPDR = 0x24000000, IDR = 0x4000, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
(gdb) x 0x40010008
0x40010008:	0x00000000
(gdb) x 0x4001000C
0x4001000c:	0x00000000
(gdb) x 0x40010010
0x40010010:	0x00000000
(gdb) x 0x40010014
0x40010014:	0x00000000

```
I break run to see EXTI CR after configuration phase and still 0..
```
(gdb) x 0x40010008
0x40010008:	0x00000000
(gdb) x 0x4001000C
0x4001000c:	0x00000000
(gdb) x 0x40010010
0x40010010:	0x00000000
(gdb) x 0x40010014
0x40010014:	0x00000000
```

cybertale commented:
@erwango  Odd, nucleo_l073rz's button is connected with pin 13 of gpioc, which base address should be 0x50000800, while result of your gpio is 0x50000000, which is the base address of GPIOA.

erwango commented:
> @erwango Odd, nucleo_l073rz's button is connected with pin 13 of gpioc, which base address should be 0x50000800, while result of your gpio is 0x50000000, which is the base address of GPIOA.

Button samples enable UART So this is PA2 configuration... Need to test again

erwango commented:
Beginning of function
```
Breakpoint 1, stm32_gpio_configure (base_addr=0x50000800, pin=pin@entry=13, conf=0, altf=altf@entry=0) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:149
149		pin = 1 << pin;
(gdb) p gpio
$1 = (GPIO_TypeDef *) 0x50000800
(gdb) p pin
$2 = 13
(gdb) p /x (GPIO_TypeDef)*gpio
$3 = {MODER = 0xffffffff, OTYPER = 0x0, OSPEEDR = 0x0, PUPDR = 0x0, IDR = 0x0, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
```
End of function:
```
Breakpoint 2, stm32_gpio_configure (base_addr=0x50000800, pin=8192, pin@entry=13, conf=0, altf=altf@entry=0) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:202
202		return 0;
(gdb) p /x (GPIO_TypeDef)*gpio
$4 = {MODER = 0xf3ffffff, OTYPER = 0x0, OSPEEDR = 0x0, PUPDR = 0x0, IDR = 0x2000, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
```

cybertale commented:
@erwango Good! GPIO configuration seems fine, what about the EXTICRs?

cybertale commented:
I'm not sure when the interrupt is enabled, I usually break at the reads of the port, interrupt should be initialized that time.

erwango commented:
> I'm not sure when the interrupt is enabled, I usually break at the reads of the port, interrupt should be initialized that time.

I break after init. For F4 I can see config is ok. For L0, everything is 00

erwango commented:
Btw, I'll be off for almost 2 weeks in few hours. I won't be able to help anymore during that time.

cybertale commented:
OK, then I'd better take a good use of these hours! :)
Let me have a look at the code.

cybertale commented:
Then it should be the problem of EXTI, could you help break at soc_gpio.c:258 and print port, pin, line? I think this is where problems should be.

erwango commented:
```
Breakpoint 1, stm32_gpio_enable_int (port=2, pin=pin@entry=13) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:260
260		if (port == STM32_PORTH) {
(gdb) p pin
$1 = 13
(gdb) p port
$2 = 2
(gdb) p /x line
$4 = 0xf00003

```

cybertale commented:
Sorry, I didn't notice that there are difference in LL for l0 and f4, I found the problem.

cybertale commented:
@erwango I fixed the problem, please give it a try if you got some time.

cybertale commented:
I just change soc_gpio.c:243 from 
`line = (0xF << ((pin % 4 * 4) + 16)) | (pin / 4);`
to
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32F0X)
	line = ((pin % 4 * 4) << 16) | (pin / 4);
#else
	line = (0xF << ((pin % 4 * 4) + 16)) | (pin / 4);
#endif
```

erwango commented:
Still KO

cybertale commented:
Could you break at LL_SYSCFG_SetEXTISource(port, line) in stm32_gpio_enable_int and print out the pin and line?

erwango commented:
Working when using the lines you provided, I had an issue when collecting the update change. My bad

cybertale commented:
@erwango Great! My first experience  of remote debugging, happy it's working now. I'm thinking of merging all functions in soc_gpio.c into gpio_stm32.c, what do you think? I'm planning clean up the code as you mentioned now since you'll soon be off.

erwango commented:
Great ! 
One remaining issue: UART Ko on STM32F3....

>  I'm thinking of merging all functions in soc_gpio.c into gpio_stm32.c

This makes perfect sense

cybertale commented:
I'm thinking breaking after init, and print out the layout of gpio like you used:
`(gdb)` p /x (GPIO_TypeDef)*gpio`
and EXTICRs.
For F3, EXTICRs are still 
 0x40010008
 0x4001000c
 0x40010010
 0x40010014

erwango commented:
> For F3, EXTICRs are still

I've did a quick test and EXTI are fine for F3, blinky as well. Only issue it UART.

erwango commented:
Ok, strange.
This is working with nucleo_f334r8/nucleo_f302r8.
But on stm32f3_disco uart is not working with your change (while ok on master).

cybertale commented:
Is EXTI fine on stm32f3_disco?

cybertale commented:
UART problem should be related to afrl and afrh...I think

erwango commented:
> Is EXTI fine on stm32f3_disco?

Yes, I tested with modified sample, and it's working.
Should be UART pin config issue

cybertale commented:
It's better to see the full gpio registers layout of stm32f3_disco.

cybertale commented:
I mean after init, what afr[0]/[1] and moder of it contains.

erwango commented:
So console is UART1, using GPIOC pins:
```
(gdb) p /x *0x48000800  (MODER)
$2 = 0xa00000
(gdb) p /x *0x48000820 (AFRL)
$3 = 0x0
(gdb) p /x *0x48000824 (AFRH)
$4 = 0x7700
```

cybertale commented:
So AFRH should be end with 770 not 7700 if I'm calculating this right.
pin 9/10 should be using AF 7, right?

erwango commented:
> pin 9/10 should be using AF 7, right?

Exact

erwango commented:
Wait: it's actually:
```
	{STM32_PIN_PC4, STM32F3_PINMUX_FUNC_PC4_USART1_TX},
	{STM32_PIN_PC5, STM32F3_PINMUX_FUNC_PC5_USART1_RX},
```

erwango commented:
Issue is in stm32_gpio_configure at following line 
```
		if (pin < 8) {
```
pin is not pin anymore it's 1 << pin.

erwango commented:
tested ok with  a quick fix.
I'm going now!

cybertale commented:
Oh, sorry it's my fault! Great it's working!

cybertale commented:
Have a good trip! I'll try to clean up the code and manage to take care the thing you mentioned in another thread while you're away.

erwango commented on drivers/gpio/Kconfig.stm32:
```diff
@@ -9,6 +9,7 @@ menuconfig GPIO_STM32
 	bool "GPIO Driver for STM32 family of MCUs"
 	depends on GPIO && SOC_FAMILY_STM32
 	select HAS_DTS_GPIO
+	select USE_STM32_LL_GPIO
```
You don't actually need that for now.
This is made to compile the ll .c files which are not used in the drivers for now (only functions defined in ll header files are used).

erwango reviewed (changes requested):
Tested ok on all stm32 series available in Zephyr today.
One small fix and we're good!
Thanks again for the change, this saves a lot of code to maintain and even saves few cycles at execution.

cybertale commented:
Hi @erwango ,I removed it, how about this time?

cybertale commented:
And I didn't quit understand what should I do with #12177. Power operations are additions above my code, so there's nothing I should do, right?

erwango commented:
Ok, it just depends which one gets merged first. Let me ack this one and we'll see

erwango commented:
ok, seems a rebase is required first

erwango commented:
@cybertale , required rebase is due to commit https://github.com/zephyrproject-rtos/zephyr/commit/359adfc9598758c5d35965473abd897813e709e2 from https://github.com/zephyrproject-rtos/zephyr/pull/11667 which is applying MISRA-C rule about boolean operands evaluation. Since you're deleting the files, it doesn't matter much but please take these rule into account in the code you're adding.

cybertale committed:
soc: arm: st_stm32: Cleanup gpio function code.

All series STM32 have mostly the same GPIO architecture
and can share the same code for GPIO manipulation.
Functions of the external interrupt line control are also the same.
This patch extracts common code from them and put them into the 'common'
folder.

Functions of control GPIO of these series scattered in
soc/arm/st_stm32/stm32xx/ folders contain these functions:
stm32_gpio_flags_to_conf(), stm32_gpio_configure(), stm32_gpio_set(),
stm32_gpio_get, stm32_gpio_enable_int().
This patch merges them into the gpio_stm32.c file.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
soc: arm: st_stm32: Using LL library to implement gpio functions

The original implementation of gpio functions access registers
directly. Using LL library can add a set of unifying access
functions for all series of stm32 for avoiding accessing low level
code, and improve readability.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale commented:
Hi @erwango, I updated the code, please give it a look if you got some time!

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -37,6 +36,226 @@ static void gpio_stm32_isr(int line, void *arg)
+	} else {
+		LL_GPIO_SetPinMode(gpio, pin, LL_GPIO_MODE_ALTERNATE);
+
+		if (pin < (1 << 8)) {
```
This is correct but confusing. 
We should introduce an alternate variable for pin used with LL api  (pin_ll ? Anything better is welcome) so we could use a simple and clear if statement here (if pin < 8).

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -54,12 +273,12 @@ static int gpio_stm32_config(struct device *dev, int access_op,
 	}
 
-	if (stm32_gpio_configure(cfg->base, pin, pincfg, 0)) {
+	if (gpio_stm32_configure(cfg->base, pin, pincfg, 0)) {
```
Add a "!= 0" to the if statement

erwango reviewed (changes requested):
Last comments and we should be good.

cybertale commented:
Hi @erwango , I added some '!=0' and '== 0' to some if in gpio_stm32.c and the additional pin_ll variable, how about this time?

erwango reviewed (approved):
Tested on 10 boards (at least one from each supported STM32 series)
